File: Basic Statistics_v2.py
2 + 2 # Function F9
# works as calculator

# Python Libraries (Packages)
# pip install <package name> - To install library (package), execute the code in Command prompt
# pip install pandas

import pandas as pd

# Read data into Python
education = pd.read_csv("C:/Data/education.csv")

# C:\Users\education.csv - this is windows default file path with a '\'
# C:\\Users\\education.csv - change it to '\\' to make it work in Python

#Exploratory Data Analysis
#Measures of Central Tendency / First moment business decision
education.workex.mean() # '.' is used to refer to the variables within object
education.workex.median()
education.workex.mode()

from scipy import stats
stats.mode(education.workex)

# Measures of Dispersion / Second moment business decision
education.workex.var() # variance
education.workex.std() #standard deviation
range = max(education.workex) - min(education.workex) # range
range

# Third moment business decision
education.workex.skew()

#Fourth moment business decision
education.workex.kurt()

#Graphical Representation
import matplotlib.pyplot as plt # mostly used for visualization purposes 
import numpy as np

plt.bar(height = education.gmat, x = np.arange(1,774,1)) # initializing the parameter

plt.hist(education.gmat) #histogram

plt.boxplot(education.gmat) #boxplot


#Normal Quantile-Quantile Plot
import scipy.stats as stats
import pylab

# Checking Whether data is normally distributed
stats.probplot(education.gmat, dist="norm",plot=pylab)

stats.probplot(education.workex,dist="norm",plot=pylab)

#transformation to make workex variable normal
import numpy as np
stats.probplot(np.log(education.workex),dist="norm",plot=pylab)

# z-distribution
# cdf => cumulative distributive function; # similar to pnorm in R
stats.norm.cdf(740,711,29)  # given a value, find the probability

# ppf => Percent point function; # similar to qnorm in R
stats.norm.ppf(0.975,0,1) # given probability, find the Z value

#t-distribution
stats.t.cdf(1.98,139) # given a value, find the probability; # similar to pt in R
stats.t.ppf(0.975, 139) # given probability, find the t value; # similar to qt in R


##############################################################
###### Data Preprocessing########################################

## import packages
import pandas as pd
import matplotlib.pyplot as plt 
import numpy as np
import seaborn as sns

##################  creating Dummy variables using dummies ###############
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# we use ethinc diversity dataset  for this
df = pd.read_csv("C:/Users/abdul/Downloads/ethnic diversity.csv")


df.columns
# drop emp_name column
df.drop(['Employee_Name','EmpID','Zip'],axis =1,inplace =True)
df.dtypes


######################################
# Create dummy variables on categorcal columns

df_new = pd.get_dummies(df)



### we have created dummies for all categorical columns

#######lets us see using one hot encoding works

df1 = pd.read_csv("C:/Users/abdul/Downloads/ethnic diversity.csv")
df1.drop(['Employee_Name', 'EmpID','Zip'], axis =1, inplace =True)
from sklearn.preprocessing import OneHotEncoder
# creating instance of one-hot-encoder
enc = OneHotEncoder(handle_unknown='ignore')

enc_df = pd.DataFrame(enc.fit_transform(df1).toarray())

############################################################
########/\/\/\/\/\/\\\/\/\/\//\/\\/\\\/\/\/\/\\\####################

df2 = pd.read_csv('C:/Users/abdul/Downloads/ethnic diversity.csv')
df2.drop(['Employee_Name', 'EmpID','Zip'], axis =1, inplace =True)


from sklearn.preprocessing import LabelEncoder
# creating instance of labelencoder
labelencoder = LabelEncoder()
X = df2.iloc[:, 0:9]
y = df2['Race']

df2.columns

X['Sex']= labelencoder.fit_transform(X['Sex'])
X['MaritalDesc'] = labelencoder.fit_transform(X['MaritalDesc'])
X['CitizenDesc'] = labelencoder.fit_transform(X['CitizenDesc'])


########## label encode y

y = labelencoder.fit_transform(y)
y = pd.DataFrame(y)

### we have to convert y to data frame so that we can use concatenate function

# concatenate X and y

df_new = pd.concat([X, y], axis =1)
## rename column name
df_new.columns
df_new = df_new.rename(columns={0:'Type'})
#################################################################################


import pandas as pd
import matplotlib.pyplot as plt 
import numpy as np
import seaborn as sns

## load data set

ethnic= pd.read_csv("C:/Users/abdul/Downloads/ethnic diversity.csv")

ethnic.columns


ethnic.isna().sum()
ethnic.isnull().sum()


# Normalization function using z std. all are continuous data.
def norm_func(i):
    x = (i-i.mean())/(i.std())
    return (x)

# Normalized data frame (considering the numerical part of data)
df_norm = norm_func(ethnic.iloc[:,10:12])
df_norm.describe()

##################################################################################
###################### Outlier Treatment ###################################
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

df = pd.read_csv("C:/Users/abdul/Downloads/ethnic diversity.csv") # for doing modifications
df.dtypes
df.isna().sum()

# let's find outliers in RM
sns.boxplot(df.Salaries);plt.title('Boxplot');plt.show() #

sns.boxplot(df.age);plt.title('Boxplot');plt.show() #


# Detection of outliers (find limits for RM based on IQR)
IQR = df['Salaries'].quantile(0.75) - df['Salaries'].quantile(0.25)
lower_limit = df['Salaries'].quantile(0.25) - (IQR * 1.5)
upper_limit = df['Salaries'].quantile(0.75) + (IQR * 1.5)



############### 1. Remove (let's trimm the dataset) ################
# Trimming Technique
# let's flag the outliers in the data set
outliers_df = np.where(df['Salaries'] > upper_limit, True, np.where(df['Salaries'] < lower_limit, True, False))
df_trimmed = df.loc[~(outliers_df), ]
df.shape, df_trimmed.shape

# let's explore outliers in the trimmed dataset
sns.boxplot(df_trimmed.Salaries);plt.title('Boxplot');plt.show()

#we see no outiers


####################### 2.Replace ############################
# Now let's replace the outliers by the maximum and minimum limit
df['df_replaced']= pd.DataFrame(np.where(df['Salaries'] > upper_limit, upper_limit, 
                                         np.where(df['Salaries'] < lower_limit, lower_limit, df['Salaries'])))
sns.boxplot(df.df_replaced);plt.title('Boxplot');plt.show()

###################### 3. Winsorization #####################################
from feature_engine.outlier_removers import Winsorizer
windsoriser = Winsorizer(distribution='skewed', # choose skewed for IQR rule boundaries or gaussian for mean and std
                          tail='both', # cap left, right or both tails 
                          fold=1.5,
                          variables=['Salaries'])
df_t = windsoriser.fit_transform(df[['Salaries']])

# we can inspect the minimum caps and maximum caps 
windsoriser.left_tail_caps_, windsoriser.right_tail_caps_

# lets see boxplot
sns.boxplot(df_t.Salaries);plt.title('Boxplot');plt.show()
###################################################################################
#################### Missing Values Imputation ##################################
import numpy as np
import pandas as pd

# load the dataset
# use modified ethnic dataset
df_raw = pd.read_csv('C:/Users/abdul/Downloads/modifiedethnic diversity.csv') # raw data without doing any modifications
df = pd.read_csv('C:/Users/abdul/Downloads/modifiedethnic diversity.csv') # for doing modifications

# check for count of NA'sin each column
df.isna().sum()

# There are 3 columns that have missing data ---Create an imputer object that fills 'Nan' values of SEX,MaritalDesc,Salaries
# Mean and Median imputer are used for numeric data (Salaries)
# mode is used for discrete data (SEX,MaritalDesc) 

# for Mean,Meadian,Mode imputation we can use Simple Imputer or df.fillna()
from sklearn.impute import SimpleImputer

# Mean Imputer 
mean_imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
df["Salaries"] = pd.DataFrame(mean_imputer.fit_transform(df[["Salaries"]]))
df["Salaries"].isnull().sum()  # all 2 records replaced by mean 

# Median Imputer
median_imputer = SimpleImputer(missing_values=np.nan, strategy='median')
df["Salaries"] = pd.DataFrame(median_imputer.fit_transform(df[["Salaries"]]))
df["Salaries"].isnull().sum()  # all 2 records replaced by median 

# Mode Imputer
mode_imputer = SimpleImputer(missing_values=np.nan, strategy='most_frequent')
df["Sex"] = pd.DataFrame(mode_imputer.fit_transform(df[["Sex"]]))
df["MaritalDesc"] = pd.DataFrame(mode_imputer.fit_transform(df[["MaritalDesc"]]))
df.isnull().sum()  # all SEX,MaritalDesc,Salaries records replaced by mode

##############################################################################
################## Type casting###############################################
import pandas as pd

data = pd.read_csv("C:/Users/abdul/Downloads/ethnic diversity.csv")
data.dtypes

#type casting
# Now we will convert 'float64' into 'int64' type. 
data.Salaries = data.Salaries.astype('int64') 
data.dtypes


#Identify duplicates records in the data
duplicate = data.duplicated()
sum(duplicate)

#Removing Duplicates
data1 = data.drop_duplicates() 

##############################################################################
#################### String manipulation ######################################

word = "Keep Adapting"
print(word)

#Accessing

word = "Keep Adapting"
letter=word[0]

print(letter)

#length 
word = "Keep Adapting"

len(word)

#finding

word = "Keep Adapting"
print(word.count('p')) # count how many times p is in the string
print(word.find("t")) # find the word t in the string
print(word.index("Adapting")) # find the letters Adapting in the string

s = "The world won't care about your self-esteem. The world will expect you to accomplish something BEFORE you feel good about yourself."

print(s.count(' '))

#Slicing

word1 = "_$_the internet frees us from the responsibility of having to retain anything in our long-term memory@_."

print (word1[0]) #get one char of the word
print (word1[0:1]) #get one char of the word (same as above)
print (word1[0:3]) #get the first three char
print (word1[:3]) #get the first three char
print (word1[-3:]) #get the last three char
print (word1[3:]) #get all but the three first char
print (word1[:-3]) #get all but the three last character
print (word1[3:-3]) #get all 


# spliting

word3 = 'Good health is not something we can buy. However, it can be an extremely valuable savings account.'

word3.split(' ') # Split on whitespace
['Good','health','is','not','something','we','can','buy.','However,','it','can','be','an','extremely','valuable','savings','account.']

# Startswith / Endswith
word3 = 'Remain calm, because peace equals power.'
word3.startswith("R")
word3.endswith("d")
word3.endswith("c")

# repeat string 

print( " & "* 108 )# prints ten dots


# replacing

word4 = "Live Happily"

word4.replace("Live", "Lead Life")



# Reversing
string = "eert a tsniaga pu mih deit yehT .meht htiw noil eht koot dna tserof eht otni emac sretnuh wef a ,yad enO "

print (''.join(reversed(string)))

#Strip
#Python strings have the strip(), lstrip(), rstrip() methods for removing
#any character from both ends of a string.

#If the characters to be removed are not specified then white-space will be removed

File: Basic Statistics_v2.r
2 + 2 # control + enter  or control + R
# works as calculator

#Packages
install.packages("readr")
library(readr)

install.packages()

# Read data into R
education <- read_csv("C:/Data/education.csv")
education <- read.csv(file.choose()) # load csv file into R

# C:\Users\\education.csv - this is windows default file path with a '\'
# C:\\Users\\Desktop\\education.csv - change it to '\\' to make it work in R

View(education)

#Exploratory Data Analysis
#Measures of Central Tendency / First moment business decision

mean(education$workex) # '$' is used to refer to the variables within object
attach(education) # When used we can directly refer to the variable name

mean(gmat)
mean(workex)

rm(xyz) #Remove specific object to free RAM (memory)
rm(list=ls()) # Remove all to free RAM (memory)

median(workex) #Median

mode(workex)

#Mode
y <- c(19, 4, 5, 7, 29, 19,19,19,19, 29, 13, 25,29,5,5,5,5,4,4,4,4)
# uni mode function
Mode <- function(x){
     a = unique(x) # x is a vector
     return(a[which.max(tabulate(match(x, a)))])
}

Mode(y)

# Bi mode function 
Modes <- function(x) {
  ux <- unique(x)
  tab <- tabulate(match(x, ux))
  ux[tab == max(tab)]
}
Modes(y)

# Measures of Dispersion / Second moment business decision
var(workex) # variance
sd(workex) #standard deviation
range <- max(workex) - min(workex) # range

install.packages("moments")
library(moments)

#Third moment business decision
skewness(workex)

#Fourth moment business decision
kurtosis(workex)

#Graphical Representation
barplot(gmat)
dotchart(gmat)

hist(gmat) #histogram

boxplot(gmat) #boxplot
y <- boxplot(gmat)
y$out # to see outliers

# Probability Distribution
install.packages("UsingR")
library("UsingR")
densityplot(gmat)

# Normal Quantile-Quantile Plot
qqnorm(gmat)
qqline(gmat)
qqnorm(workex)
qqline(workex)

# Transformation to make workex variable normal
qqnorm(log(workex))
qqline(log(workex))



#### Data Pre-processing ##############################

## Dummy variables creation 
# use ethnic diversity dataset
# packages such as 'dummyvars', 'fastdummies' can be used 
# loading the dataset
data <- read.csv(file.choose())

# checking str and summary of the data
str(data)
summary(data)
attach(data)

install.packages("fastDummies")
library(fastDummies)

data_dummy <- dummy_cols(data, select_columns = c("Position","State","Sex","MaritalDesc","CitizenDesc","EmploymentStatus","Department","Race"),
                      remove_first_dummy = TRUE,remove_most_frequent_dummy = FALSE,remove_selected_columns = TRUE)
??fastDummies
# you can choose between remove_first_dummy and remove_most_frequent_dummy
# as both will remove multicollinearity issue 
#################################################
### Label encoding in R
library(CatEncoders)
View(data)

# character column: 'Position'
lb_new <- LabelEncoder.fit(data$Position)
# new values are transformed to NA
position_new<- transform(lb_new,data$Position)
position_new

# Using cbind to combine with original dataset
newdata <- cbind(data, position_new)

################################################
###### Normalization ###################################
# we will ethnic diversity dataset
df <- data_dummy[, -c(1,2,3)]

# to normalize the data we use custom function 
norm <- function(x){
  return ((x-min(x))/(max(x)-min(x)))
}

df_norm <- as.data.frame(lapply(df, norm))
# Normalized data have values ranging between 0 to 1

# To apply standardization we have inbuilt function scale
# we use mtcars dataset 
df_mt <- read.csv(file.choose())
View(df_mt)
# remove first column 
df_mt <- df_mt[,-1]
# use scale function 
df_scale <- as.data.frame(scale(df_mt))
# hypothetically you have values ranging between -3 to +3 or in general the range is -inf to +inf

###########//\\\\//////\\\\\\\\/////\\\\\\////\\\\\\///\\\\\//\\\\##############

# Handling duplicates###############################################
# Duplicate entries are removed using 'duplicated' function
# it stores the duplicate values into another variable

dup <- duplicated(data)
dup
data_new <- data[!duplicated(data), ]
data_new

####### /\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/ ####################
###################### Type Casting ########################
# Type casting is used to convert one data type into another.
# ex: categorical data to numeric data

as.numeric(data$age) 

is.numeric(data$age) # it will return true or false

# Ex: Convert Numeric to Integer
data$age <- as.integer(data$age)

# Ex: Convert categorical data to factor type
str(data) # categorical data is character

# convert sex column to factors
data$Sex <- as.factor(data$Sex)
str(data) # check now for Sex column

# Alternatively we can use argument stringAsFactors=TRUE
# load ethnic diversity dataset
data1 <- read.csv(file.choose(),stringsAsFactors = TRUE)
str(data1)
summary(data1)

################################################
## Missing values - imputation

# Lets introduce NA values into the dataset using 'missForest' package

install.packages("missForest")
library(missForest)

# Generate 10% missing values at Random
# remove three unwanted columns as they are not required for analysis
data.mis <- prodNA(data[,-c(1,2,5)], noNA = 0.1)
summary(data.mis)
attach(data.mis)
sum(is.na(data.mis)) # 310

# Omitting NA values from the Data 
data_omit <- na.omit(data.mis) # na.omit => will omit the rows which has atleast 1 NA value
dim(data_omit)
sum(is.na(data_omit)) #0

# Alternatively We can apply mean/median/mode imputation
data_new <- data.mis # work with original data for imputation
summary(data_new)

# NA values are present in both continuous and categorical columns

# Mean imputation for continuous data - CLMAGE
data_new$Salaries[is.na(data_new$Salaries)] <- mean(data_new$Salaries, na.rm = TRUE)
data_new$age[is.na(data_new$age)] <- mean(data_new$age, na.rm = TRUE)


# Mode imputation for categorical data
# Custom function to calculate Mode
Mode <- function(x){
  a = table(x) # x is a vector
  names(a[which.max(a)])
}

data_new$Position[is.na(data_new$Position)] <- Mode(data_new$Position[!is.na(data_new$Position)])

# select categorical columns with NA, so we simple remove columns which 
# does not have NA and use apply function to impute all columns at once
data_cat <- data_new[,-c(1,8,9)]
sum(is.na(data_cat)) # 210
# create a new dataframe to use apply function 
# apply is inbuilt function which takes arguments as
### X ---- dataframe, matrix, list
#### MARGIN --- 2 implies to columns and 1 implies to rows
#### FUN ---- Implies to functions suchas mean, sum, average, or any custom defined function
data_tran <- as.data.frame(apply(data_cat,c(1,2),Mode))
sum(is.na(data_tran)) # 0

# now combine using cbind function
clean_data <- cbind(data_new[,c(1,8,9)],data_tran)
sum(is.na(clean_data)) 

######################################################################
##########/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\############3
# Outlier Treatment
data<- read.csv(file.choose())
sum(is.na(data))
attach(data)
View(data)
boxplot(data$Salaries)
boxplot(data$age)

# There are 2 outliers in the salary column
# Replace outliers with the maximum value - Winsorization
qunt1 <- quantile(data$Salaries,probs = c(.25,.75))
qunt1 # 25% 23092, # 75% = 51452
caps <- quantile(data$Salaries, probs = c(.01,.99), na.rm = T)
caps # 1 % = 441, 99% = 95770
H <- 1.5*IQR( data$Salaries,na.rm = T)
H # 42539.92
data$Salaries[data$Salaries<(qunt1[1]-H)] <- caps[1]
data$Salaries[data$Salaries>(qunt1[2]+H)] <- caps[2]
boxplot(data$Salaries)


####################################################
############# zero Variance
library(ggplot2)
library(ggthemes)
# Use 'apply' and 'var' functions to 
# check for variance on numerical values
apply(data, 2, var)

# Check for the columns having zero variance

which(apply(data, 2, var)==0) # ignore the warnings


###################
# z-distribution
pnorm(680,711,29) # given a value, find the probability
qnorm(0.025) # given probability, find the Z value


# t-distribution
pt(1.98, 139) # given a value, find the probability
qt(0.975, 139) # given probability, find the t value
